Post Editor reduxify

jsnajdr opened the pull request:
This branch contains a very big diff with completely reduxified post editor and with no Flux-based `PostEditStore` at all. Still has many bugs, but can be built and is good enough for testing.

Will be landed as one big PR or, preferrably, split into multiple standalone PRs.

jsnajdr committed:
Post Editor: reduxify

jsnajdr committed:
Leave post.meta.data property when normalizing post for state

Remove only `meta.links` fields. The purpose of this removal is to save
space in local storage. See also #7141.

jsnajdr committed:
cleanup new post init

nylen commented on client/post-editor/editor-featured-image/preview-container.jsx:
```diff
@@ -84,9 +83,9 @@ export default class extends React.Component {
-				} );
+				// PostActions.edit( {
+				// 	featured_image: image.ID,
+				// } );
```
Remove rather than comment out?

jsnajdr replied:
Thanks @nylen for the reminder about this "todo". I converted this Flux dispatch to a Redux one in https://github.com/Automattic/wp-calypso/pull/25160/commits/b7d8e28126f138276c2d3fa1c90493219e529739.

The dispatch is necessary for several cases to work correctly:
- uploading a new image to media library and choosing it as featured img before the upload is complete. This ensures that the transient ID is updated to the permanent one.
- setting featured image by selecting it from Google library. Again, transient ID vs permanent

The Google library case was actually reported as an issue recently in #24807. I forgot to update this one from Flux to Redux when migrating the Featured Image component. I verified that the patch fixes the issue.

alisterscott commented:
I'm investigating why so many e2e tests are failing and I will leave separate comments for each issue

alisterscott commented:
Clicking "Edit" after Publish breaks the editor (and this is causing most of the editor e2e tests to fail):

![edit](https://user-images.githubusercontent.com/128826/40756868-03013b1c-64c9-11e8-9917-4dc785ae6294.gif)

alisterscott commented:
The trash post e2e test fails with a similar issue, but I can't reproduce this one manually, maybe because it performs the actions quite quickly?

This is what it looks like when it fails (after clicking the Trash Post button):

<img width="1109" alt="trash post" src="https://user-images.githubusercontent.com/128826/40757052-f8a1ebb6-64c9-11e8-9561-83b95c02f8f8.png">

jsnajdr commented:
Thanks @alisterscott for investigating the e2e failures. Most of them are caused by weird `page.show` and `page.replace` calls when the editor transitions between certain states. I'm debugging them right now.

jsnajdr committed:
Fix redirects after saving a new draft and incorrect post type

Move the `page.replace` calls away from lifecycle methods: they were performed
also on editor open/close transitions. Perform them in the `onSaveSuccess` method,
the only place where it can ever happen.

jsnajdr commented:
@alisterscott It seems that all e2e failures had a common cause and got magically fixed with a single commit.

There are still 4 failures reported, but they don't seem to be related to Post Editor:
<img width="950" alt="screen shot 2018-05-31 at 15 18 39" src="https://user-images.githubusercontent.com/664258/40784510-27867bb0-64e6-11e8-8bd8-498fd3e5e16c.png">

jsnajdr committed:
Don't copy post tags: they are already in terms

jsnajdr committed:
Pass saveResult to onSaveDraftSuccess also on autosave

Fixes a bug where the URL didn't get updated to include post ID after it was
assigned on first save of a new draft.

jsnajdr commented:
The full e2e suite is green :tada:

alisterscott commented:
> There are still 4 failures reported, but they don't seem to be related to Post Editor:

Yes, looking at the [screenshots](https://circleci.com/gh/Automattic/wp-e2e-tests-for-branches/8391#artifacts/containers/0) it looks like domains were having some issues which are since resolved

jsnajdr committed:
Post Editor: don't fail when autosaved post is not dirty

Also, little refactoring of `isPublished` check, and re-introduce setting
the `content` in saved attributes when only raw content differs.

jsnajdr committed:
Post Editor: fix client/state tests

jsnajdr committed:
Post Editor: recover from null post in ContactFormDialog

jsnajdr committed:
Post Editor: avoid back and forth flickering when switching editor modes

Store editor mode in component state and use the editor preferences only as
initial value. Avoids flickering of the visual/html toggle:

1. Click "Visual", UI switched, preference save request with 'tinymce' value
2. Click "HTML", UI switched, preference save request with 'html' value
3. 1st preference save finishes, response contains 'tinymce', UI switched to 'tinymce' (!)
4. 2nd preference save finishes, response contains 'html', UI switched to 'html' (!)

The last two switches are unwanted, are an artefact of the preference's local and remote
value and resetting them after successful save request.

jsnajdr committed:
Post Editor: correctly reset raw content after mode switch

Wait for `setState( { mode } )` to be performed -- only then the TinyMCE is guaranteed
to be switched.
Next, save the content to post edits and reset the raw content to the value after switch
(Visual and HTML modes differ in present and missing HTML markup for the same content)
so that `initial === current`.

jsnajdr committed:
Post Editor: remove extra status edit

jsnajdr committed:
Post Editor: save featured image ID change after transient media item becomes permanent

Convert a (commented) Flux action dispatch to Redux dispatch, and move the call up the
component hierarchy to universal `onImageChange` prop. Fixes #24807.

jsnajdr committed:
Post Editor: edit content in Redux only when it's changed

We sync content from TinyMCE to Redux before save and after editor mode switch. Create
the Redux edit only when content has really changed and not on every save/switch.

jsnajdr committed:
Remove outdated Flux post edit store tests

jsnajdr committed:
Fix lib/media/actions tests

jsnajdr committed:
Fix EditorWordCount tests after raw content was reduxified

jsnajdr committed:
Post Editor: fix utils return values and tests

jsnajdr committed:
Remove unneeded mocks from posts/utils tests

jsnajdr committed:
Post Editor: move selecting media item to Simple Payments onImageChange handler

Move the media-item-selecting code to Simple Payments' `onImageChange` handler where it
originally was. Don't call it for feature image: that causes weird race conditions in e2e tests.

jsnajdr commented:
Current status: all unit and e2e tests are 💚, all issues from call for testing are fixed and no new ones are coming. I can't find any more bugs myself either. Going to merge on Tuesday morning (CET) after alerting happiness to watch for post editor issues.

blowery reviewed (approved):
LGTM! Great work.

apeatling commented:
💯
